File: s3PythonSDK/sample/ObjectSerialTesting.py
import sys
from botocore.exceptions import ClientError
from client import client

# test 1.Basic put bucket
#      2.put Object with header & custom metadata
#      3.head object
#      4.get object
#      5.copy object (with preserved ACL)


def percent_cb(complete, total):
    sys.stdout.write('.')
    sys.stdout.flush()


def main(arg):
    try:
        content = "abcdefghijklmnopqrstuvwxyz<br>01234567890112345678901234<br>!@#$%^&*()-=[]{};':',.<>/?<br>01234567890112345678901234<br>abcdefghijklmnopqrstuvwxyz<br>"

        client.create_bucket(
            CreateBucketConfiguration={'LocationConstraint': 'ap-northeast-1'},
            Bucket=arg[0],
        )
        client.create_bucket(
            CreateBucketConfiguration={'LocationConstraint': 'ap-northeast-1'},
            Bucket=arg[1],
        )
        print("Put Object 'test.txt' with Headers & custom metadata..")

        client.put_object(
            Body=content,
            ACL='public-read',
            Bucket=arg[0],
            CacheControl='no-cache',
            ContentDisposition='attachment;filename=\'default.txt\'',
            ContentEncoding='UTF-8',
            ContentType='text/plain',
            Key='test.txt',
            Metadata={
                'x-amz-meta-flower': 'lily',
                'x-amz-meta-color': 'pink'
            },
        )
        response = client.get_object(
            Bucket=arg[0],
            Key='test.txt',
        )
        #print("Cache-Control: " + repr(response['ResponseMetadata']['HTTPHeaders']['cache-control']))
        # print("Content-Type"+repr(response['ResponseMetadata']['HTTPHeaders']['content-type']))
        #print("Content-Encoding: "+repr(response['ResponseMetadata']['HTTPHeaders']['content-encoding']))
        #print("Content-Disposition: "+repr(response['ResponseMetadata']['HTTPHeaders']['content-disposition']))
        #print("x-amz-meta-x-amz-meta-flower: "+repr(response['ResponseMetadata']['HTTPHeaders']['x-amz-meta-x-amz-meta-flower']))
        #print("x-amz-meta-x-amz-meta-color: "+repr(response['ResponseMetadata']['HTTPHeaders']['x-amz-meta-x-amz-meta-color']))
        #print("\nCopy yuyuman1://test.txt to yuyuman2 as 'cptest' and preserve_acl..")
        url = client.generate_presigned_url(
            'get_object',
            Params={
                'Bucket': arg[0],
                'Key': 'test.txt'
            },
            ExpiresIn=120,
        )
        # print(url)
        client.copy_object(
            ACL='public-read',
            Bucket=arg[1],
            CopySource=arg[0]+'/test.txt',
            Key='cptest',
        )
        response = client.get_object(
            Bucket=arg[1],
            Key='cptest',
        )
        #print("\nGet file content as string..\n    " +response['Body'].read().decode("utf-8") )
        fp = open('download.test', 'wb+')
        client.download_fileobj(arg[0], 'test.txt', fp)
        client.download_file(arg[0], 'test.txt', 'download.test2')
        client.upload_fileobj(fp, arg[0], 'test.txt')
        fp.close()
        f2 = open('testwrite', 'wb+')
        client.upload_fileobj(f2, arg[0], 'test.txt')
        f2.close()
        client.upload_file('download.test2', arg[0], 'test.txt')
        #print("\nClean up..")
        # clear bucket
        result = client.list_objects_v2(
            Bucket=arg[0],
        )
        if 'Contents' in result:
            for r in result['Contents']:
                client.delete_object(
                    Bucket=arg[0],
                    Key=r['Key']
                )
        client.delete_bucket(
            Bucket=arg[0],
        )
        result = client.list_objects_v2(
            Bucket=arg[1],
        )
        if 'Contents' in result:
            for r in result['Contents']:
                client.delete_object(
                    Bucket=arg[1],
                    Key=r['Key']
                )
        client.delete_bucket(
            Bucket=arg[1],
        )

        print("Object Serial Test done!\n")
    except ClientError as e:
        print("Error operation : " + e.operation_name)
        print("Error response : " + e.response['Error']['Message'])

File: s3PhpSDKExample/sample/ObjectSerialTesting.php

<?php
require 'client.php';

date_default_timezone_set('Asia/Taipei');

function createSampleFile()
{
    $temp = tmpfile();
    $content = "abcdefghijklmnopqrstuvwxyz<br>01234567890112345678901234<br>!@#$%^&*()-=[]{};':',.<>/?<br>01234567890112345678901234<br>abcdefghijklmnopqrstuvwxyz<br>";
    fwrite($temp, $content);
    fseek($temp, 0);
    return $temp;
    fclose($temp); // this removes the file
}
echo "Object Serial Testing...\n";
$bucketname=$argv[1];
$bucketname2=$argv[2];
$objName="test.txt";
$objName2="test2.txt";
$datetime = new DateTime('17 Oct 2100');
$datetime2 = new DateTime('17 Oct 2019');

$client->createBucket(array(
    'Bucket' => $bucketname
));
$client->createBucket(array(
    'Bucket' => $bucketname2
));

$client->putObject(array(
    'Bucket' => $bucketname,
    'Key'    => $objName,
    'Body'   => createSampleFile(),
    'Expires' => "GMT ".$datetime->format('c'),
));

$result=$client->getObject(array(
    'Bucket' => $bucketname,
        'Key'=> $objName
));

$client->deleteObject(array(
    'Bucket' => $bucketname,
    'Key'	 => $objName
));


//put with acl & custom meta
$client->putObject(array(
        'Bucket' => $bucketname,
        'Key'    => $objName,
        'Body'   => createSampleFile(),
        'ACL'	 => 'public-read',
        'Metadata' => array(
                'x-amz-meta-flower' => 'lily',
                'x-amz-meta-color' => "pink"
        ),
        'ContentType' => "text/plain",
        'ContentLength' => '150',
        'ContentDisposition'=> "attachment; filename=\"default.txt\"",
        'ContentMD5'=>'movf4FeaK/4LQyz5FP1oiQ=='
));

$result=$client->getObject(array(
        'Bucket' => $bucketname,
        'Key'=> $objName
));
$a=str_replace('"', "", $result['ETag']);
//echo $a;
$client->copyObject(array(
        'Bucket' => $bucketname2,
        'Key'    => $objName2,
        'Body'   => createSampleFile(),
        'ACL'	 => 'public-read-write',
        'CopySource'=> $bucketname.'/'.$objName,
        'MetadataDirective' => 'REPLACE',
        'Metadata' => array(
                'x-amz-meta-flower' => 'lily',
                'x-amz-meta-color' => "pink"
        ),
        'ContentType' => "image/jpeg",
));

$client->copyObject(array(
        'Bucket' => $bucketname2,
        'Key'    => 'test4',
        'Body'   => createSampleFile(),
        'ACL'	 => 'public-read-write',
        'CopySource'=> $bucketname.'/'.$objName,
        'Expires' => "GMT ".$datetime->format('c'),
        

));
$client->copyObject(array(
        'Bucket' => $bucketname2,
        'Key'    => 'test5',
        'Body'   => createSampleFile(),
        'ACL'	 => 'public-read-write',
        'CopySource'=> $bucketname.'/'.$objName,
        'CopySourceIfmodifiedSince' =>  "GMT ".$datetime2->format('c')


));
$client->copyObject(array(
        'Bucket' => $bucketname2,
        'Key'    => 'test6',
        'Body'   => createSampleFile(),
        'ACL'	 => 'public-read-write',
        'CopySource'=> $bucketname.'/'.$objName,
        'CopySourceIfUnmodifiedSince' =>  "GMT ".$datetime->format('c')
));

$client->copyObject(array(
        'Bucket' => $bucketname2,
        'Key'    => 'test7',
        'Body'   => createSampleFile(),
        'ACL'	 => 'public-read-write',
        'CopySource'=> $bucketname.'/'.$objName,
        'MetadataDirective' => 'COPY',
        'WebsiteRedirectLocation' => 'http://www.google.com'


));
// Head Object

$client->headObject(array(
        'Bucket' => $bucketname,
        'Key'=> $objName,
        'IfMatch' => $a
));
$client->headObject(array(
        'Bucket' => $bucketname,
        'Key'=> $objName,
        'IfNoneMatch' => '123'
));
$client->headObject(array(
        'Bucket' => $bucketname,
        'Key'=> $objName,
        'IfModifiedSince' => "GMT ".$datetime2->format('c')
));
$client->headObject(array(
        'Bucket' => $bucketname,
        'Key'=> $objName,
        'IfUnmodifiedSince' => "GMT ".$datetime->format('c')
));
$client->headObject(array(
        'Bucket' => $bucketname,
        'Key'=> $objName,
        'Range' => 'bytes=50-100'
));


// Get Object
$client->getObject(array(
		'Bucket' => $bucketname,
		'Key'=> $objName,
		'IfMatch' => $a,
		'Range' => 'bytes=50-100',
		'ResponseContentDisposition' => 'attachment; filename="default.txt"',
		'ResponseContentLanguage' => 'en',
		'ResponseContentType' => 'text/plain',
		'ResponseExpires' => "GMT ".$datetime->format('c'),
));

$client->getObject(array(
		'Bucket' => $bucketname,
		'Key'=> $objName,
		'IfNoneMatch' => '123'
));

$client->getObject(array(
		'Bucket' => $bucketname,
		'Key'=> $objName,
		'IfModifiedSince' => "GMT ".$datetime2->format('c')
));

$client->getObject(array(
		'Bucket' => $bucketname,
		'Key'=> $objName,
		'IfUnmodifiedSince' => "GMT ".$datetime->format('c')
));

$client->getObject(array(
		'Bucket' => $bucketname,
		'Key'=> $objName,
		'Range' => 'bytes=0-100',
		'SaveAs' => '123.456'
));


$client->deleteObject(array(
		'Bucket' => $bucketname,
		'Key'	 => $objName
));
$client->deleteObject(array(
		'Bucket' => $bucketname2,
		'Key'	 => 'test4'
));
$client->deleteObject(array(
		'Bucket' => $bucketname2,
		'Key'	 => 'test5'
));
$client->deleteObject(array(
		'Bucket' => $bucketname2,
		'Key'	 => 'test6'
));
$client->deleteObject(array(
		'Bucket' => $bucketname2,
		'Key'	 => 'test7'
));

$client->deleteObject(array(
		'Bucket' => $bucketname2,
		'Key'	 => $objName2
));
$client->deleteBucket(array(
    	'Bucket' => $bucketname
));
$client->deleteBucket(array(
    	'Bucket' => $bucketname2
));
